### PR TITLE
publish button with options

### LINKED ISSUE #409: Multiple publish options
There should be two options for publishing minutes:
- For `University Network` and `Student`
- For `Student` only

### PR DESCRIPTION
**Pull request by Nef10**
fix #409

### CODE REVIEW

File: _1327/documents/urls.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 	url(r"(?P<title>[\w-]+)/edit$", views.edit, name='edit'),
 	url(r"(?P<title>[\w-]+)/autosave$", views.autosave, name='autosave'),
 	url(r"(?P<title>[\w-]+)/publish$", views.publish, name='publish'),
+	url(r"(?P<title>[\w-]+)/publish_student$", views.publish_student, name='publish_student'),
```
Comment (Bartzi) [inline _1327/documents/urls.py]: wouldn't it be better to do this using only one url endpoint and distinguishing between students and university network using a POST/GET param?

**Comment by Nef10** [comment]
@Bartzi please have another look

File: _1327/documents/views.py [hunk 2]
```diff
@@ -177,14 +177,14 @@ def permissions(request, title):
...
+def publish(request, title, state_id):
 	document = get_object_or_404(Document, url_title=title)
 	check_permissions(document, request.user, [document.edit_permission_name])
 	if not document.show_publish_button():
 		raise PermissionDenied()
 
-	document.publish()
-	messages.success(request, _("Minutes document has been published."))
+	document.publish(state_id)
+	messages.success(request, _("Minutes document has been published for Students and University Network."))
```
Comment (janno42) [inline _1327/documents/views.py]: Please differentiate

File: _1327/minutes/models.py [hunk 3]
```diff
@@ -40,11 +41,13 @@ class MinutesDocument(Document):
...
 	INTERNAL = 2
 	CUSTOM = 3
+	PUBLISHED_STUDENT = 4
 	CHOICES = (
 		(UNPUBLISHED, _('Unpublished')),
-		(PUBLISHED, _('Published')),
+		(PUBLISHED, _('Published for Students and University Network')),
 		(INTERNAL, _('Internal')),
 		(CUSTOM, _('Custom')),
+		(PUBLISHED_STUDENT, _('Published for Student only')),
```
Comment (janno42) [inline _1327/minutes/models.py]: Students

File: _1327/minutes/models.py [hunk 4]
```diff
@@ -123,12 +129,16 @@ def handle_edit(self, cleaned_data):
...
+	student_group = Group.objects.get(name=settings.STUDENT_GROUP_NAME)
+	university_network_group = Group.objects.get(name=settings.UNIVERSITY_GROUP_NAME)
+	if instance.state == MinutesDocument.UNPUBLISHED or instance.state == MinutesDocument.INTERNAL:
+		instance.delete_all_permissions(student_group)
+		instance.delete_all_permissions(university_network_group)
+	if instance.state == MinutesDocument.PUBLISHED:
+		assign_perm(instance.view_permission_name, student_group, instance)
+		assign_perm(instance.view_permission_name, university_network_group, instance)
+	if instance.state == MinutesDocument.PUBLISHED_STUDENT:
+		assign_perm(instance.view_permission_name, student_group, instance)
```
Comment (janno42) [inline _1327/minutes/models.py]: Remove permissions for university network here

**Comment by Nef10** [comment]
@janno42 please have another look

File: _1327/minutes/migrations/0007_auto_20160927_1937.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('minutes', '0006_auto_20160927_1615'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='minutesdocument',
+            name='state',
+            field=models.IntegerField(choices=[(0, 'Unpublished'), (1, 'Published for Students and University Network'), (2, 'Internal'), (3, 'Custom'), (4, 'Published for Students only')], default=0, verbose_name='State'),
```
Comment (janno42) [inline _1327/minutes/migrations/0007_auto_20160927_1937.py]: Students only

File: _1327/minutes/models.py [hunk 6]
```diff
@@ -92,9 +95,12 @@ def show_permissions_editor(self):
 	def show_publish_button(self):
 		return self.state == MinutesDocument.UNPUBLISHED
 
-	def publish(self):
-		self.state = MinutesDocument.PUBLISHED
-		self.save()
+	def publish(self, state_id):
+		if state_id not in (MinutesDocument.PUBLISHED, MinutesDocument.PUBLISHED_STUDENT):
```
Comment (Bartzi) [inline _1327/minutes/models.py]: I guess you wanted to write `if state_id in ...`

File: _1327/minutes/tests.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def test_publish_button(self):
 		document = mommy.make(MinutesDocument, participants=self.participants, moderator=self.moderator, state=MinutesDocument.UNPUBLISHED)
 		document.set_all_permissions(staff_group)
 
-		self.app.get(reverse('documents:publish', args=[document.url_title]), user=self.user)
+		self.app.get(reverse('documents:publish', args=[document.url_title, 1]), user=self.user)
```
Comment (Bartzi) [inline _1327/minutes/tests.py]: Could you write `staff_group.id` instead of `1`

File: _1327/minutes/tests.py [hunk 8]
```diff
@@ -82,6 +82,43 @@ def test_publish_button(self):
...
+	def test_publish_student_button(self):
+		"""
+		Test if the publish for students only button works
+		"""
+		staff_group = Group.objects.get(name=settings.STAFF_GROUP_NAME)
+
+		document = mommy.make(MinutesDocument, participants=self.participants, moderator=self.moderator, state=MinutesDocument.UNPUBLISHED)
+		document.set_all_permissions(staff_group)
+
+		self.app.get(reverse('documents:publish', args=[document.url_title, 4]), user=self.user)
```
Comment (Bartzi) [inline _1327/minutes/tests.py]: Same applies here

File: _1327/minutes/models.py (hunk 6 above)
Comment (janno42) [inline _1327/minutes/models.py]: Only the two states he lists there are allowed. So `not` is correct.

Comment (janno42) [inline _1327/minutes/models.py]: Nevermind, the code snippet was not helpful to review. I think you are right :)

File: _1327/minutes/tests.py (hunk 7 above)
Comment (janno42) [inline _1327/minutes/tests.py]: This is the state id not a group id. The states are not easily accessible in the template so hard coding is okay.

Comment (janno42) [inline _1327/minutes/tests.py]: But a comment would be nice

Comment (Bartzi) [inline _1327/minutes/tests.py]: Yeah right ;)
Comment should be sufficient then

**Pull request merged**